File: src/easy/Merge_Sorted_Array.java
package easy;

import java.util.Arrays;

/**
 * https://leetcode.com/problems/merge-sorted-array/
 *
 * You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.
 *
 * Merge nums1 and nums2 into a single array sorted in non-decreasing order.
 *
 * The final sorted array should not be returned by the function, but instead be stored inside the array nums1. To accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that should be merged, and the last n elements are set to 0 and should be ignored. nums2 has a length of n.\
 *
 * TC: O((m+n)log(m+n))
 * SC: O(1)
 *
 */
public class Merge_Sorted_Array {
    class Solution {
        public void merge(int[] nums1, int m, int[] nums2, int n) {
            int index = 0;
            for(int i = m; i < nums1.length; i++) {
                nums1[i] = nums2[index++];
            }

            Arrays.sort(nums1);
        }
    }
}

File: src/easy/merge_sorted_array.cpp
class Solution {
public:
//method 1: simple solution using STL sort 
//TC:O((m+n)log(m+n))
//SC:O(1)
    
    
//     void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        
//         for(int i=m; i<m+n; i++)
//             nums1[i]+=nums2[i-m];
    
//         sort(nums1.begin(), nums1.end());
//     }
    
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        
//method 2: simple solution using the fact that two given arrays are sorted.
//TC: O(m+n)
//SC: O(1)
        
        int i=m-1, j=n-1, end=m+n-1;
        
        while(end>=0)
        {
            if(i==-1)
                nums1[end]=nums2[j--];
            else if(j==-1)
                nums1[end]==nums1[i--];
            else if(nums1[i]>nums2[j])
                nums1[end]=nums1[i--];
            else
                nums1[end]=nums2[j--];
            end--;
        }
        
    }
};
